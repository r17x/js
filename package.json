{
	"name": "js",
	"private": true,
	"engines": {
		"node": ">=20"
	},
	"workspaces": ["packages/*"],
	"author": "r17x <hi@rin.rocks>",
	"license": "MIT",
	"packageManager": "yarn@4.5.0",
	"scripts": {
		"prepare": "husky",
		"release": "yarn exec multi-semantic-release",
		"test": "vitest",
		"build": "yarn workspaces foreach -Rpt --since --no-private run build",
		"lint": "run-s \"lint:*\"",
		"lint:biome": "biome ci .",
		"lint:test": "echo \"No lint:test script specified\""
	},
	"lint-staged": {
		"*": [
			"biome check --no-errors-on-unmatched --files-ignore-unknown=true --write"
		]
	},
	"dependencies": {
		"@anolilab/multi-semantic-release": "^1.1.3",
		"@biomejs/biome": "^1.9.2",
		"@commitlint/cli": "^19.5.0",
		"@commitlint/config-conventional": "^19.5.0",
		"@vitest/coverage-v8": "^2",
		"commitlint-config-squash-pr": "^1.1.0",
		"husky": "^9.1.6",
		"lint-staged": "^15.2.10",
		"npm-run-all": "^4.1.5",
		"semantic-release": "^24.1.1",
		"vitest": "^2"
	}
}
