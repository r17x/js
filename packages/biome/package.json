{
	"name": "@r17x/biome",
	"description": "biomejs.dev configuration",
	"version": "1.0.2",
	"homepage": "https://github.com/r17x/js/tree/main/packages/biome",
	"repository": {
		"type": "git",
		"url": "git+https://github.com/r17x/js.git",
		"directory": "packages/biome"
	},
	"publishConfig": {
		"access": "public",
		"provenance": true
	},
	"type": "module",
	"files": [
		"index.json"
	],
	"exports": {
		".": "./index.json"
	}
}
