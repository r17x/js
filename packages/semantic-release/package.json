{
  "name": "@r17x/semantic-release",
  "description": "semantic-release configuration",
  "version": "1.0.0",
  "homepage": "https://github.com/r17x/js/tree/main/packages/semantic-release",
  "repository": {
    "type": "git",
    "url": "https://github.com/r17x/js.git",
    "directory": "packages/semantic-release"
  },
  "publishConfig": {
    "access": "public",
    "provenance": true
  },
  "files": [
    "index.json"
  ],
  "exports": {
    ".": {
      "require": "./index.json",
      "import": "./index.json"
    }
  },
  "dependencies": {
    "@semantic-release/changelog": "^6.0.3",
    "@semantic-release/commit-analyzer": "^13.0.0",
    "@semantic-release/exec": "^6.0.3",
    "@semantic-release/git": "^10.0.1",
    "@semantic-release/github": "^10.3.4",
    "@semantic-release/npm": "^12.0.1",
    "@semantic-release/release-notes-generator": "^14.0.1",
    "conventional-changelog-conventionalcommits": "^8.0.0"
  }
}
