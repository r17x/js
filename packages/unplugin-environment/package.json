{
	"name": "unplugin-environment",
	"type": "module",
	"version": "1.1.0",
	"description": "A plugin for loading enviroment variables safely with schema validation, simple with virtual module, type-safe with intellisense, and better DX ðŸ”¥ ðŸš€ ðŸ‘·. Support with Next.js, Vite, Webpack, Rollup, Rspack, Farm, and more.",
	"license": "MIT",
	"author": "r17x <hi@rin.rocks>",
	"homepage": "https://github.com/r17x/js#readme",
	"repository": {
		"type": "git",
		"url": "https://github.com/r17x/js.git",
		"directory": "packages/unplugin-environment"
	},
	"bugs": {
		"url": "https://github.com/r17x/js/issues"
	},
	"keywords": [
		"plugins",
		"unplugin",
		"vite",
		"webpack",
		"rollup",
		"transform",
		"farm",
		"environment",
		"env",
		"dotenv",
		"te-env"
	],
	"exports": {
		".": {
			"types": "./dist/index.d.ts",
			"import": "./dist/index.js",
			"require": "./dist/index.cjs"
		},
		"./astro": {
			"types": "./dist/astro.d.ts",
			"import": "./dist/astro.js",
			"require": "./dist/astro.cjs"
		},
		"./rspack": {
			"types": "./dist/rspack.d.ts",
			"import": "./dist/rspack.js",
			"require": "./dist/rspack.cjs"
		},
		"./farm": {
			"types": "./dist/farm.d.ts",
			"import": "./dist/farm.js",
			"require": "./dist/farm.cjs"
		},
		"./vite": {
			"types": "./dist/vite.d.ts",
			"import": "./dist/vite.js",
			"require": "./dist/vite.cjs"
		},
		"./webpack": {
			"types": "./dist/webpack.d.ts",
			"import": "./dist/webpack.js",
			"require": "./dist/webpack.cjs"
		},
		"./rollup": {
			"types": "./dist/rollup.d.ts",
			"import": "./dist/rollup.js",
			"require": "./dist/rollup.cjs"
		},
		"./esbuild": {
			"types": "./dist/esbuild.d.ts",
			"import": "./dist/esbuild.js",
			"require": "./dist/esbuild.cjs"
		},
		"./types": {
			"types": "./dist/types.d.ts",
			"import": "./dist/types.js",
			"require": "./dist/types.cjs"
		},
		"./*": "./*"
	},
	"main": "dist/index.cjs",
	"module": "dist/index.js",
	"types": "dist/index.d.ts",
	"typesVersions": {
		"*": {
			"*": ["./dist/*", "./*"]
		}
	},
	"files": ["dist", "client.d.ts"],
	"scripts": {
		"prepack": "npm run build",
		"build": "tsup",
		"build:fix": "esno ../../tooling/unplugin/postbuild.ts",
		"dev": "tsup --watch src"
	},
	"peerDependencies": {
		"@farmfe/core": ">=1",
		"astro": ">=4",
		"esbuild": "*",
		"rollup": "^3",
		"vite": ">=3",
		"webpack": "^4 || ^5"
	},
	"peerDependenciesMeta": {
		"@farmfe/core": {
			"optional": true
		},
		"esbuild": {
			"optional": true
		},
		"rollup": {
			"optional": true
		},
		"vite": {
			"optional": true
		},
		"webpack": {
			"optional": true
		}
	},
	"dependencies": {
		"@mobily/ts-belt": "4.0.0-rc.1",
		"@rollup/pluginutils": "^5.1.0",
		"dotenv": "^16.4.5",
		"npm-run-path": "^5.3.0",
		"picocolors": "^1.0.1",
		"typescript": "^5.3.2",
		"unplugin": "1.10.1",
		"zod": "^3.23.8",
		"zod-to-ts": "^1.2.0"
	},
	"devDependencies": {
		"astro": "^4.11.1",
		"bumpp": "^9.2.0",
		"chalk": "^5.3.0",
		"eslint": "^8.55.0",
		"esno": "^4.0.0",
		"fast-glob": "^3.3.2",
		"nodemon": "^3.0.2",
		"npm-run-all": "^4.1.5",
		"rollup": "^4.6.1",
		"tsup": "^8.0.1",
		"vite": "^5.0.4",
		"webpack": "^5.89.0"
	}
}
