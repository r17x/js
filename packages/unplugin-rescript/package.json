{
	"name": "unplugin-rescript",
	"type": "module",
	"version": "0.0.2",
	"description": "a plugin for integration ReScript with most Bundler (Vite, Rollup, Webpack, Esbuild, etc).",
	"license": "MIT",
	"author": "r17x <hi@rin.rocks>",
	"homepage": "https://github.com/r17x/js#readme",
	"repository": {
		"type": "git",
		"url": "https://github.com/r17x/js.git",
		"directory": "packages/unplugin-rescript"
	},
	"bugs": {
		"url": "https://github.com/r17x/unplugin-rescript/issues"
	},
	"keywords": [
		"plugins",
		"unplugin",
		"vite",
		"webpack",
		"rollup",
		"transform",
		"farm",
		"rescript"
	],
	"exports": {
		".": {
			"types": "./dist/index.d.ts",
			"import": "./dist/index.js",
			"require": "./dist/index.cjs"
		},
		"./astro": {
			"types": "./dist/astro.d.ts",
			"import": "./dist/astro.js",
			"require": "./dist/astro.cjs"
		},
		"./rspack": {
			"types": "./dist/rspack.d.ts",
			"import": "./dist/rspack.js",
			"require": "./dist/rspack.cjs"
		},
		"./farm": {
			"types": "./dist/farm.d.ts",
			"import": "./dist/farm.js",
			"require": "./dist/farm.cjs"
		},
		"./vite": {
			"types": "./dist/vite.d.ts",
			"import": "./dist/vite.js",
			"require": "./dist/vite.cjs"
		},
		"./webpack": {
			"types": "./dist/webpack.d.ts",
			"import": "./dist/webpack.js",
			"require": "./dist/webpack.cjs"
		},
		"./rollup": {
			"types": "./dist/rollup.d.ts",
			"import": "./dist/rollup.js",
			"require": "./dist/rollup.cjs"
		},
		"./esbuild": {
			"types": "./dist/esbuild.d.ts",
			"import": "./dist/esbuild.js",
			"require": "./dist/esbuild.cjs"
		},
		"./types": {
			"types": "./dist/types.d.ts",
			"import": "./dist/types.js",
			"require": "./dist/types.cjs"
		},
		"./*": "./*"
	},
	"main": "dist/index.cjs",
	"module": "dist/index.js",
	"types": "dist/index.d.ts",
	"typesVersions": {
		"*": {
			"*": ["./dist/*", "./*"]
		}
	},
	"files": ["dist"],
	"scripts": {
		"prepack": "npm run build",
		"build": "tsup",
		"build:fix": "esno ../../tooling/unplugin/postbuild.ts",
		"dev": "tsup --watch src"
	},
	"peerDependencies": {
		"@farmfe/core": ">=1",
		"astro": ">=4",
		"esbuild": "*",
		"rollup": "^3",
		"vite": ">=3",
		"webpack": "^4 || ^5"
	},
	"peerDependenciesMeta": {
		"@farmfe/core": {
			"optional": true
		},
		"esbuild": {
			"optional": true
		},
		"rollup": {
			"optional": true
		},
		"vite": {
			"optional": true
		},
		"webpack": {
			"optional": true
		}
	},
	"dependencies": {
		"@rollup/pluginutils": "^5.1.0",
		"npm-run-path": "^5.3.0",
		"picocolors": "^1.0.1",
		"unplugin": "1.10.1"
	},
	"devDependencies": {
		"astro": "^4.11.1",
		"bumpp": "^9.2.0",
		"chalk": "^5.3.0",
		"eslint": "^8.55.0",
		"esno": "^4.0.0",
		"fast-glob": "^3.3.2",
		"nodemon": "^3.0.2",
		"npm-run-all": "^4.1.5",
		"rollup": "^4.6.1",
		"tsup": "^8.0.1",
		"typescript": "^5.3.2",
		"vite": "^5.0.4",
		"webpack": "^5.89.0"
	}
}
