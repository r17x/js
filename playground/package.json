{
	"name": "playground",
	"private": true,
	"scripts": {
		"clean": "rescript clean -with-deps",
		"dev:vite": "vite dev",
		"dev:farm": "farm dev",
		"build:vite": "vite build",
		"build:farm": "farm build"
	},
	"devDependencies": {
		"@farmfe/cli": "^1.0.2",
		"@farmfe/core": "^1.2.0",
		"@rescript/core": "1.5.0",
		"@rescript/react": "^0.12.2",
		"rescript": "11.1.1",
		"vite": "^5.0.4",
		"vite-plugin-inspect": "^0.8.1"
	},
	"dependencies": {
		"react": "^18.3.1",
		"react-dom": "^18.3.1"
	}
}
