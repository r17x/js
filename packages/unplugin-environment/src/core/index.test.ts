import { beforeEach, describe, expect, it, vi } from "vitest";
import { z } from "zod";
import { zodToTs } from "zod-to-ts";
import * as Core from ".";
import { name as pkgName, version as pkgVersion } from "../../package.json";
import type { Options, PluginOption } from "./types";

describe("uplugin-environment:core", () => {
	beforeEach(vi.clearAllMocks);

	const casesSchema = [
		{
			schema: "",
			expected: "Error",
		},
		{
			schema: z.object({
				REACT_APP_NAME: z.string(),
			}),
			expected: z.ZodObject,
		},
		{
			schema: {
				REACT_APP_NAME: z.string(),
			},
			expected: z.ZodObject,
		},
	];

	it.each(casesSchema)(
		"$schema field of PluginOption is $expected",
		({ schema: s, expected }) => {
			const { data, success } = Core.schema.safeParse(s);
			if (expected === "Error") {
				expect(success).toBe(false);
			} else {
				expect(success).toBe(true);
				expect(data).instanceof(expected);
			}
		},
	);

	const casesOptions = [
		{
			options: "APP",
			expected: {
				match: "APP",
				schema: expect.any(z.ZodType),
				moduleEnvName: "@env",
			},
		},
		{
			options: ["APP", "!SERVER"],
			expected: {
				match: ["APP", "!SERVER"],
				schema: expect.any(z.ZodType),
				moduleEnvName: "@env",
			},
		},
		{
			options: {
				match: ["APP", "!SERVER"],
				schema: {
					APP_NAME: z.string(),
				},
			},
			expected: {
				match: ["APP", "!SERVER"],
				schema: expect.any(z.ZodType),
				moduleEnvName: "@env",
			},
		},
	];

	it.each(casesOptions)(
		"Valid $options as param of Environment",
		({ options, expected }) => {
			expect(Core.getOptions(options)).toStrictEqual(expected);
		},
	);

	it("valid merge zod schema", () => {
		expect(
			Core.printTypeDefinition(
				zodToTs(
					z
						.object({
							REACT_APP_NAME: z.string().optional(),
						})
						.merge(
							z.object({
								REACT_APP_VERSION: z.string(),
							}),
						),
				).node,
			),
		).matchSnapshot();
	});

	it("only run watchChange once when valid watchList", () => {
		const onChange = vi.fn();
		const watcher = Core.watchChange([".env"], onChange);
		watcher("/a/b/c/.env");
		watcher("/a/b/c/.env~~~~");
		watcher("/a/b/c/any");
		watcher("/a/b/c/any");
		watcher("/a/b/c/any");
		expect(onChange).toHaveBeenCalledTimes(1);
	});

	it("only for knowledge", () => {
		expect(
			Core.printTypeDefinition(
				zodToTs(z.record(z.any().optional(), z.string())).node,
			),
		).toMatchSnapshot();

		expect(
			Core.printTypeDefinition(
				zodToTs(
					z.object({
						REACT_APP_NAME: z.string(),
						REACT_APP_VERSION: z.string(),
						REACT_APP_AGE: z.number(),
					}),
				).node,
			),
		).toMatchSnapshot();
	});

	it("[DTS] valid create module type definition", () => {
		const options = Core.getOptions({
			match: "REACT_APP",
			schema: {
				REACT_APP_NAME: z.string(),
				REACT_APP_VERSION: z.string(),
				REACT_APP_AGE: z.number(),
				REACT_APP_PORT: z.coerce.number().min(1),
				REACT_APP_DATA: z
					.string()
					.transform((a) => JSON.parse(a))
					.pipe(
						z.object({
							age: z.number(),
							name: z.string(),
						}),
					),
			},
			moduleEnvName: "@myenv",
		});

		expect(
			Core.createModuleDTS(
				{
					REACT_APP_NAME: "htmx_BTW",
					REACT_APP_VERSION: "6.6.6",
					REACT_APP_DURATION: "123",
					REACT_APP_PORT: "8080",
					REACT_APP_DATA: JSON.stringify({ age: 123, name: "htmx_BTW" }),
					SECRET_KEY_TOKEN_: "1234567890",
				},
				options as Options,
			),
		).toMatchInlineSnapshot(
			`
"// DO NOT EDIT THIS FILE. IT IS GENERATED BY \`${pkgName}@${pkgVersion}\` - see https://npm.im/unplugin-environment

declare module "@myenv" {
export const env: {
    REACT_APP_NAME: string;
    REACT_APP_VERSION: string;
    REACT_APP_DURATION?: string;
    REACT_APP_PORT: number;
    REACT_APP_DATA: {
        age: number;
        name: string;
    };
    REACT_APP_AGE: number;
}
}"
`,
		);
	});

	it("print type definition", () => {
		expect(
			Core.printTypeDefinition(
				zodToTs(
					z
						.object({
							REACT_APP_NAME: z.string(),
						})
						.strict(),
				).node,
			),
		).toBe(
			`{
    REACT_APP_NAME: string;
}`,
		);
	});

	it("valid parse env", () => {
		const mockEnv = {
			REACT_APP_PORT: "8080",
			REACT_APP_NAME: "htmx_BTW",
			REACT_APP_VERSION: "6.6.6",
			SECRET_KEY_TOKEN_: "1234567890",
		};

		expect(
			Core.getEnv(mockEnv, Core.getOptions("REACT_APP") as Options),
		).toStrictEqual({
			REACT_APP_NAME: "htmx_BTW",
			REACT_APP_VERSION: "6.6.6",
			REACT_APP_PORT: "8080",
		});

		expect(
			Core.getEnv(
				mockEnv,
				Core.getOptions(["REACT_APP", "!REACT_APP_VERSION"]) as Options,
			),
		).toStrictEqual({
			REACT_APP_NAME: "htmx_BTW",
			REACT_APP_PORT: "8080",
		});

		expect(
			Core.getEnv(
				mockEnv,
				Core.getOptions({
					match: ["REACT_APP", "!REACT_APP_VERSION"],
					schema: z.object({
						REACT_APP_NAME: z.string(),
						REACT_APP_PORT: z.coerce.number().min(1),
					}),
				}) as Options,
			),
		).toStrictEqual({
			REACT_APP_NAME: "htmx_BTW",
			REACT_APP_PORT: 8080,
		});
	});

	it("valid resolveId", () => {
		const options = Core.getOptions("REACT_APP");
		Core.mapOptions(options, {
			single: (o) => {
				expect(Core.loadInclude(o)("@env")).toBeTruthy();
				expect(Core.loadInclude(o)("@env.ts")).toBeFalsy();
				expect(Core.loadInclude(o)("@env.ts")).toBeFalsy();
			},
			clientServer: () => {},
		});
	});

	it("valid options value", () => {
		expect(() => Core.getOptions("")).toThrowError();
		expect(() => Core.getOptions({} as PluginOption)).toThrowError();
		expect(() => Core.getOptions({ match: "" } as PluginOption)).toThrowError();
		expect(() =>
			Core.getOptions({ client: "", server: "" } as PluginOption),
		).toThrowError();

		expect(Core.getOptions("REACT_APP")).toStrictEqual({
			match: "REACT_APP",
			schema: expect.any(z.ZodType),
			moduleEnvName: "@env",
		});

		expect(Core.getOptions(["REACT_APP", "VITE_APP"])).toStrictEqual({
			match: ["REACT_APP", "VITE_APP"],
			schema: expect.any(z.ZodType),
			moduleEnvName: "@env",
		});

		expect(
			Core.getOptions({ client: "CLIENT_APP", server: "SERVER_APP" }),
		).toStrictEqual({
			client: {
				match: "CLIENT_APP",
				schema: expect.any(z.ZodType),
				moduleEnvName: "@env/client",
			},
			server: {
				match: "SERVER_APP",
				schema: expect.any(z.ZodType),
				moduleEnvName: "@env/server",
			},
		});

		expect(
			Core.getOptions({
				client: {
					match: "CLIENT_APP",
					schema: z.object({
						CLIENT_APP_NAME: z.string(),
					}),
					moduleEnvName: "@myenv/client",
				},
				server: {
					match: "SERVER_APP",
					schema: z.object({
						SERVER_APP_NAME: z.string(),
					}),
					moduleEnvName: "@myenv/server",
				},
			}),
		).toStrictEqual({
			client: {
				match: "CLIENT_APP",
				schema: expect.any(z.ZodType),
				moduleEnvName: "@myenv/client",
			},
			server: {
				match: "SERVER_APP",
				schema: expect.any(z.ZodType),
				moduleEnvName: "@myenv/server",
			},
		});
	});
});
