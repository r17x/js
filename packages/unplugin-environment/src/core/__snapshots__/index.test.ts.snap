// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`uplugin-environment:core > only for knowledge 1`] = `
"{
    [x: string]: string;
}"
`;

exports[`uplugin-environment:core > only for knowledge 2`] = `
"{
    REACT_APP_NAME: string;
    REACT_APP_VERSION: string;
    REACT_APP_AGE: number;
}"
`;

exports[`uplugin-environment:core > valid merge zod schema 1`] = `
"{
    REACT_APP_NAME?: string | undefined;
    REACT_APP_VERSION: string;
}"
`;
